{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyPoOfoInvAHREyUJwVwKG7k"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","metadata":{"id":"I48lxcjwNuSK"},"source":["# **What are Errors?**\n"," - Errors are issues in the code that prevent it from executing successfully.\n","\n","\n"," **Types of Errors**\n","  - 1. Synatx Errors\n","  - 2. Runtime Erros\n","  - 3. Logical Errors\n","\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"FWDOPGhwNwC6"},"source":["  **1. Syntax Errors**\n","   - Occur when the code vilates python's syntax rules.\n","   - Detetcted before execution.\n","   - Example: Missing colons, unmatched parentheses.\n","\n","    _   missing a colon at the end of the if statement\n","    \n","    if 5>3\n","    print(\"this willl cause a syntax error\")\n","\n","    Error Message :\n","     Syntax Error: invalid syntax\n","\n","\n","  **The python interpreter cannot undersstand the code because of incorrect syntax, such as missing colons, unmatched parentheses. or msissplled keywords.**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"pXPDWxNKNrR7","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1737695070990,"user_tz":-330,"elapsed":395,"user":{"displayName":"Gabriel Ps","userId":"06455618489830639069"}},"outputId":"90d0eba9-b2c9-4b91-fdb3-55e2b08f35ea"},"outputs":[{"output_type":"stream","name":"stdout","text":["tis will cause a syntax error\n"]}],"source":["if 5>3:\n","  print(\"tis will cause a syntax error\")"]},{"cell_type":"markdown","metadata":{"id":"wloHXroIN21i"},"source":["**2. Runtime Errors**\n","\n"," - Occur during program execution\n"," - Example: Division by zero , accessing undefined variable\n","\n","\n","\n","\n"," Example Code\n"," Division by zero :\n","\n","\n"," x=1=/0\n"," print(x)\n","\n","\n"," Error Message :\n"," zero Division Error : division by zero\n","\n","\n"," **The program syntax is correct, but the operation fails at runtime(eg, dividing by zero, accessing an undefined variable , or file not found).**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"LzQGSDr6NrOa","colab":{"base_uri":"https://localhost:8080/","height":158},"executionInfo":{"status":"error","timestamp":1737695038097,"user_tz":-330,"elapsed":7,"user":{"displayName":"Gabriel Ps","userId":"06455618489830639069"}},"outputId":"3dc01e46-3eca-4eb0-cde4-1c1f83d99a1d"},"outputs":[{"output_type":"error","ename":"ZeroDivisionError","evalue":"division by zero","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-2-5db56e641f9b>\u001b[0m in \u001b[0;36m<cell line: 0>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"]}],"source":["x=1/0\n","print(x)"]},{"cell_type":"markdown","metadata":{"id":"ySyZAkBLN9CS"},"source":["**3. Logical Errors**\n"," - The program runs but gives incorrect results due to logic flaws,\n"," - Example: incorrect formula or condition.\n","\n","\n","\n"," incorrect formula for calculating the area of a rectangle\n","\n","length =5\n","\n","width= 10\n","\n","area = length +width # should be length * width\n","\n","print(\"Area;\",area)\n","\n","ouput:\n","Area :15\n","\n","\n","\n","\n","\n","\n","**The program runs successsfully but, the result is wrong because of a mustake in the logic of formula**"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"C4zsgeyhNrLS"},"outputs":[],"source":["# incorrect formula for calculating the area of a rectangle\n","length =5\n","\n","width= 10\n","\n","area = length + width # should be length * width\n","\n","print(\"Area:\",area)"]},{"cell_type":"markdown","metadata":{"id":"m2Z_rRfFK9v9"},"source":["## What are Execptions?\n","\n","Exeptions are runtime errros that occur during program execution, disrupting the normal flow of a program\n","\n","**Common Built-in Exceptions:**\n"," - ZeroDivisionErrors\n"," - ValueError\n"," - KeyError\n"," - FileNotFoundError\n"," - indexError"]},{"cell_type":"markdown","metadata":{"id":"-yhyBiK7GtyL"},"source":["## The Try-Except Block\n","\n","**Basis Structrue:**\n","\n","(try:\n","  code that might raise an exception\n","\n","  except Esception Type:\n","    code to handle the exception)\n","\n","\n","Example code:\n","\n","try:\n","\n","x= int (input(\"Eneter a number:\"))\n","y=10/x\n","\n","exceptzeroDivisionError:\n","\n","print(\"Cannot divide by zero\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"5_UkW2Z_NrIK"},"outputs":[],"source":["try:\n","  x= int (input(\"Eneter a number:\"))\n","  y=10/x\n","except ZeroDivisionError:\n","  print(\"Cannot divide by zero\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"P2uO51_XNrEL"},"outputs":[],"source":["try:\n","  x= int (input(\"Eneter a number:\"))\n","  y=10/x\n","except ZeroDivisionError as e:\n","  print(e)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"-G5jzSaeNrAq"},"outputs":[],"source":["try:\n","  print(a) # The a is not defined and it will show error\n","except:\n","  print(\"Error occured\") # but there is exept keyword so that's why error not showing\n"]},{"cell_type":"markdown","metadata":{"id":"XXEwqO1eLDod"},"source":["If except we do, The error will not show thats the use of the try-exept block\n","\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"G6HfxBT7Nq8K"},"outputs":[],"source":["print(a)\n","print(\"Hello world\")\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"gPEFnpixNq4S"},"outputs":[],"source":["try:\n","  print(a)\n","except:\n","  print(\"Error occured\")\n","print(\"Hello world\")\n"]},{"cell_type":"markdown","metadata":{"id":"35lmF4AvLKGt"},"source":["`If except we do, The error will not show thats the use of the try-exept block`\n","`And it will not stop the code it just skip the error`\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"mEpwGJbOLMWN"},"source":["## Else and Finally Blocks\n"]},{"cell_type":"markdown","metadata":{"id":"yO-eBwu1LRss"},"source":["\n","\n","## Else Block :\n","\n","      Runs only if no esceptions occur.\n","\n","\n","Example Code :\n","\n","    try:\n","      result =10/2\n","    except ZeroDivisionError:\n","      print(\"Error\")\n","    else:\n","      print(\"Result:\", result)\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"111Wqnr5Nq0z"},"outputs":[],"source":["\n","try:\n","  result =10/2\n","except ZeroDivisionError:\n","  print(\"Error\")\n","else:\n","  print(\"Result:\", result)"]},{"cell_type":"markdown","metadata":{"id":"zFmnlGR-LXkV"},"source":["## Finally Block:\n","\n"," - Always runs, regardlesss of whether an exception occurred\n"," - Useful for cleanup (ef, closing files)\n","\n","\n"," Example Code:\n","\n","        try:\n","          file_=open(\"data.txt\",\"r\")\n","        escept FileNotFoundError:\n","          print(\"File not found!\")\n","        finally:\n","          print(\"End of program.\")\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"oQa4jfvTKw-E"},"outputs":[],"source":["try:\n","  file_= open(\"data.txt\",\"r\")\n","\n","except FileNotFoundError:\n","  print(\"File not found!\")\n","\n","finally:\n","  print(\"End of program\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"SqeCRQTmKw54"},"outputs":[],"source":["import math\n","print(math.exp(1000)) # its OverflowError"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"QgmmM2nyKw2v"},"outputs":[],"source":["obj =42\n","obj.append(5) # Attruibute Error"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"4TssZa7qKwzN"},"outputs":[],"source":["data= {\"name\":\"john\"}\n","print(data[\"age\"])  # key Error"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"c6gPomCMKww0"},"outputs":[],"source":["my_list=[1,2,3]\n","print(my_list[5]) # IndexError"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"SPLEXXWFKwtM"},"outputs":[],"source":["print(int(\"hello\")) # Value Error"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jgJw439LKwp1"},"outputs":[],"source":["my_list=[1,2,3]\n","my_list.remove(5)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"E_oNFoy9LmiO"},"outputs":[],"source":["print(\"5\"+5) #Type Error"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Ccdn3atlLmed"},"outputs":[],"source":["open(\"nonexistnt_file,.txt\")  # FileNotFoundError"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"y--e8i6lLmbc"},"outputs":[],"source":["try:\n","  open(\"nonexistnt_file.txt\")\n","except FileNotFoundError as e:\n","  print(e)"]},{"cell_type":"markdown","metadata":{"id":"RWKuUS9aLx-9"},"source":["## Raise Exception\n","\n","The raise statement in python is used to manually trigger an exception when a specific error condition occurs."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"wbgp0sJKLmXk"},"outputs":[],"source":["age = int(input(\"Enter your age:\"))\n","\n","if age<0:\n","  raise ValueError(\"Age cannot be negative\")\n","else:\n","  print(f\"Your age is{age}\")\n","\n","print(\"No try except\")\n","\n"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"jswtNQUSLmUl"},"outputs":[],"source":["try:\n","  age = int(input(\"Enter your age:\"))\n","\n","  if age<0:\n","    raise ValueError(\"Age cannot be negative\")\n","  else:\n","    print(f\"Your age is{age}\")\n","except ValueError as e:\n","  print(e)\n","print(\"try except finished\")"]},{"cell_type":"markdown","metadata":{"id":"5Z63OlBbL5v2"},"source":["## Custom Exceptions\n","\n","\n","\n","\n","Custom exceptions are user- defined exceptions in Python that help handle domain-specific errors more meaningully and improve code clarity."]},{"cell_type":"markdown","metadata":{"id":"VvZe-L4bL9Zd"},"source":["**Example Code :**\n","\n","    class CustomError(Exception)\"\n","        pass\n","    raise CustomError(\"This is a custom error\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"IKA0T3y9LmSF"},"outputs":[],"source":["class InvalidAgeException(Exception):\n","  \"Raised when the input value is less than 18\"\n","  pass\n","\n","# yout need to guess this number\n","number=18\n","try:\n","  input_num = int(input(\"Enter a number: \"))\n","  if input_num < number:\n","    raise InvalidAgeException\n","\n","  else:\n","    print(\"Eligible to Vote\")\n","\n","except InvalidAgeException:\n","  print(\"Exception Occurred : Invalid Age\")"]}]}